hooks/hookexecution/execution.go

```diff
@@ -0,0 +1,233 @@
+package hookexecution
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/prebid/prebid-server/hooks"
+	"github.com/prebid/prebid-server/hooks/hookstage"
+)
+
+type hookResponse[T any] struct {
+	Err           error
+	ExecutionTime time.Duration
+	HookID        HookID
+	Result        hookstage.HookResult[T]
+}
+
+type hookHandler[H any, P any] func(
+	context.Context,
+	hookstage.ModuleInvocationContext,
+	H,
+	P,
+) (hookstage.HookResult[P], error)
+
+func executeStage[H any, P any](
+	executionCtx executionContext,
+	plan hooks.Plan[H],
+	payload P,
+	hookHandler hookHandler[H, P],
+) (StageOutcome, P, stageModuleContext, *RejectError) {
+	stageOutcome := StageOutcome{}
+	stageOutcome.Groups = make([]GroupOutcome, 0, len(plan))
+	stageModuleCtx := stageModuleContext{}
+	stageModuleCtx.groupCtx = make([]groupModuleContext, 0, len(plan))
+
+	for _, group := range plan {
+		groupOutcome, newPayload, moduleContexts, reject := executeGroup(executionCtx, group, payload, hookHandler)
+		stageOutcome.ExecutionTimeMillis += groupOutcome.ExecutionTimeMillis
+		stageOutcome.Groups = append(stageOutcome.Groups, groupOutcome)
+		stageModuleCtx.groupCtx = append(stageModuleCtx.groupCtx, moduleContexts)
+		if reject != nil {
+			return stageOutcome, payload, stageModuleCtx, reject
+		}
+
+		payload = newPayload
+	}
+
+	return stageOutcome, payload, stageModuleCtx, nil
+}
+
+func executeGroup[H any, P any](
+	executionCtx executionContext,
+	group hooks.Group[H],
+	payload P,
+	hookHandler hookHandler[H, P],
+) (GroupOutcome, P, groupModuleContext, *RejectError) {
+	var wg sync.WaitGroup
+	done := make(chan struct{})
+	resp := make(chan hookResponse[P])
+
+	for _, hook := range group.Hooks {
+		mCtx := executionCtx.getModuleContext(hook.Module)
+		wg.Add(1)
+		go func(hw hooks.HookWrapper[H], moduleCtx hookstage.ModuleInvocationContext) {
+			defer wg.Done()
+			executeHook(moduleCtx, hw, payload, hookHandler, group.Timeout, resp, done)
+		}(hook, mCtx)
+	}
+
+	go func() {
+		wg.Wait()
+		close(resp)
+	}()
+
+	hookResponses := collectHookResponses(resp, done)
+
+	return processHookResponses(executionCtx, hookResponses, payload)
+}
+
+func executeHook[H any, P any](
+	moduleCtx hookstage.ModuleInvocationContext,
+	hw hooks.HookWrapper[H],
+	payload P,
+	hookHandler hookHandler[H, P],
+	timeout time.Duration,
+	resp chan<- hookResponse[P],
+	done <-chan struct{},
+) {
+	hookRespCh := make(chan hookResponse[P], 1)
+
+	select {
+	case <-done:
```

Comment:
This case can be removed. There is an incredibly low change a rejection would've occurred at this point where the code for this optimization isn't worth it IMHO.